map-data/src/main/java/org/triplea/map/reader/XmlParser.java

```diff
@@ -0,0 +1,137 @@
+      attributeHandlers.put(attributeName, attributeHandler);
+      return this;
+    }
+
+    public TagParser addChildTagHandler(final String tagName, final ThrowingRunnable tagHandler) {
+      return new TagParser(currentTag, attributeHandlers, new HashMap<>())
+          .addChildTagHandler(tagName, tagHandler);
+    }
+
+    public BodyParser addBodyHandler(final Consumer<String> bodyHandler) {
+      return new BodyParser(currentTag, attributeHandlers, bodyHandler);
+    }
+
+    public void parse(final XMLStreamReader xmlStreamReader) throws XMLStreamException {
+      new GenericParser(currentTag, attributeHandlers, Map.of(), body -> {}).parse(xmlStreamReader);
+    }
+  }
+
+  @AllArgsConstructor(access = AccessLevel.PRIVATE)
+  public static class TagParser {
+    private final String currentTag;
+    private final Map<String, Consumer<String>> attributeHandlers;
+    private final Map<String, ThrowingRunnable> childTagHandlers;
+
+    public TagParser addChildTagHandler(final String tagName, final ThrowingRunnable tagHandler) {
+      childTagHandlers.put(tagName, tagHandler);
+      return this;
+    }
+
+    public void parse(final XMLStreamReader streamReader) throws XMLStreamException {
+      new GenericParser(currentTag, attributeHandlers, childTagHandlers, body -> {})
+          .parse(streamReader);
+    }
+  }
+
+  @AllArgsConstructor(access = AccessLevel.PRIVATE)
+  public static class BodyParser {
+    private final String currentTag;
+    private final Map<String, Consumer<String>> attributeHandlers;
+    private Consumer<String> bodyHandler;
+
+    public void parse(final XMLStreamReader streamReader) throws XMLStreamException {
+      new GenericParser(currentTag, attributeHandlers, Map.of(), bodyHandler).parse(streamReader);
+    }
+  }
+
+  @AllArgsConstructor(access = AccessLevel.PRIVATE)
+  public static class GenericParser {
+    private final String currentTag;
+    private final Map<String, Consumer<String>> attributeHandlers;
+    private final Map<String, ThrowingRunnable> childTagHandlers;
+    private Consumer<String> bodyHandler;
+
+    public void parse(final XMLStreamReader streamReader) throws XMLStreamException {
+      boolean endTagReached = false;
+
+      if (streamReader.getEventType() == XMLStreamReader.START_ELEMENT) {
+        for (int i = 0, n = streamReader.getAttributeCount(); i < n; i++) {
+          final int index = i;
+
+          final String attributeName = streamReader.getAttributeLocalName(index);
+          final Consumer<String> attributeHandler = attributeHandlers.get(attributeName);
+          if (attributeHandler != null) {
+            String attributeValue = streamReader.getAttributeValue(index);
+            attributeHandler.accept(attributeValue);
+          }
+        }
+      }
+
+      while (streamReader.hasNext() && !endTagReached) {
+        final int event = streamReader.next();
+        switch (event) {
```

Comment:
Codacy found an issue: [Switch statements should have a default label](https://app.codacy.com/gh/triplea-game/triplea/pullRequest?prid=6102557)